File: multipaz-compose/src/commonMain/kotlin/org/multipaz/compose/cards/CardBadge.kt
package org.multipaz.compose.cards

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shadow
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.luminance
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import org.multipaz.document.DocumentBadge

/**
 * A badge to be displayed on a card.
 *
 * @property text the text to display.
 * @property color the color of the badge.
 */
data class CardBadge(
    val text: String,
    val color: Color
) {
    companion object {
        /**
         * Generates a [CardBadge] from a [DocumentBadge].
         *
         * @param badge the [DocumentBadge] to convert.
         * @return the generated [CardBadge].
         */
        fun fromDocumentBadge(badge: DocumentBadge): CardBadge {
            return CardBadge(
                text = badge.text,
                color = Color(
                    red = badge.color.red,
                    green = badge.color.green,
                    blue = badge.color.blue
                )
            )
        }
    }
}

/**
 * Renders a list of [CardBadge]s as pill-shaped rectangles in the top-right corner.
 *
 * @param badges the list of badges to render.
 * @param elevation the elevation of the badges.
 * @param modifier the modifier for the container.
 */
@Composable
fun CardBadges(
    badges: List<CardBadge>,
    elevation: Dp = 8.dp,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier.padding(12.dp),
        horizontalAlignment = Alignment.End
    ) {
        badges.forEach { badge ->
            val textColor = if (badge.color.luminance() > 0.5f) Color.Black else Color.White
            Box(
                modifier = Modifier
                    .padding(bottom = 8.dp)
                    .graphicsLayer {
                        this.shadowElevation = elevation.toPx()
                        this.shape = RoundedCornerShape(50)
                        this.clip = false
                        this.ambientShadowColor = Color.Black
                        this.spotShadowColor = Color.Black
                    }
                    .background(badge.color, RoundedCornerShape(50))
                    .border(BorderStroke(1.dp, Color.Black.copy(alpha = 0.1f)), RoundedCornerShape(50))
                    .padding(horizontal = 10.dp, vertical = 4.dp)
            ) {
                Text(
                    text = badge.text,
                    color = textColor,
                    style = MaterialTheme.typography.labelMedium.copy(
                        fontFeatureSettings = "smcp",
                        fontWeight = FontWeight.Bold,
                        shadow = Shadow(
                            color = Color.Black.copy(alpha = 0.6f),
                            offset = Offset(0f, 2f),
                            blurRadius = 4f
                        )
                    )
                )
            }
        }
    }
}

File: multipaz-swiftui/src/iosMain/swift/CardBadge.swift
import SwiftUI
import UIKit

/// A badge to be displayed on a card.
public struct CardBadge: Hashable, Sendable {
    /// the text to display.
    public let text: String
    /// the color of the badge.
    public let color: Color
    
    public init(text: String, color: Color) {
        self.text = text
        self.color = color
    }
}

/// A view that renders a list of ``CardBadge``s.
public struct CardBadgesView: View {
    let badges: [CardBadge]
    
    public init(badges: [CardBadge]) {
        self.badges = badges
    }
    
    public var body: some View {
        VStack(alignment: .trailing, spacing: 8) {
            ForEach(badges, id: \.text) { badge in
                Text(badge.text)
                    .font(Font.system(.footnote).weight(.bold).smallCaps())
                    .foregroundColor(badge.color.isLight ? .black : .white)
                    .shadow(color: .black.opacity(0.4), radius: 1, x: 0, y: 1)
                    .padding(.horizontal, 10)
                    .padding(.vertical, 4)
                    .background(
                        Capsule()
                            .fill(badge.color)
                            .shadow(color: .black.opacity(0.4), radius: 8, x: 0, y: 4)
                    )
            }
        }
        .padding(12)
    }
}

extension Color {
    var isLight: Bool {
        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0
        
        // Use UIColor to get components as SwiftUI.Color doesn't provide them easily across platforms/versions
        UIColor(self).getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        
        let luminance = 0.299 * red + 0.587 * green + 0.114 * blue
        return luminance > 0.5
    }
}
